### ISSUE TITLE
ANTLR Grammar (false) Inheritance

### ISSUE THREAD

**Issue by wiztigers** [TypeCobolTeam/TypeCobol#190]
For our `CobolCodeElements` grammar, ANTLR generates for you some source files for C# classes. Among these classes you get the class `CobolCodeElementsParser` ; this class is `public`, `partial` and lies in namespace `TypeCobol.Compiler.Parser.Generated`.
The `CobolCodeElementsParser` class defines many internal `public` `partial` class representing as many contexts as rules that are defined in the `CobolCodeElements` grammar. For example, you have `TypeCobol.Compiler.Parser.Generated.CobolCodeElementsParser.DataDescriptionEntryContext` which packs all that is needed for creating a C# AST node for the grammar rule `dataDescriptionEntry`.

We create these AST nodes by [walking the ANTLR tree](https://github.com/TypeCobolTeam/TypeCobol/blob/master/TypeCobol/Compiler/Parser/CodeElementsParserStep.cs#L165) using a [`ParseTreeWalker`](https://github.com/antlr/antlr4/blob/master/runtime/Java/src/org/antlr/v4/runtime/tree/ParseTreeWalker.java#L39L73) object. In a nice implementation of the _Visitor_ design pattern, we end up calling our context-specific methods in `CodeElementBuilder`, which do the hard stuff.

**ENTERS THE COMPOSITE GRAMMAR**

TypeCobol is essentially a dialect of COBOL. Thus, what we want is replace some, but not all, rules of our COBOL grammar with custom syntactic-sugarey rules.
So, I create a composite grammar file, like this:
``` antlr
grammar TypeCobol;

import CobolCodeElements;

// same starting rule as CodeElements
cobolCodeElements: codeElement* EOF;

// now our custom rules
dataDescriptionEntry: ... // stuff to allow strong types in addition to COBOL pictures
```
This should allow us to create a neat builder object like `CodeElementBuilder`, which would define the "enter" and "exit" methods only for the rules we explicitely override, and use `CodeElementBuilder` to do its job as it has always do.
Shouldn't be much of a problem, should it ?

**HOUSTON, WE'VE GOT A PROBLEM**

ANTLR interprets the imported grammars as if you had typed all their files content directly in the importing grammar files. In other words, it pastes text, makes a little cleanup so you don't get the same rule twice, and uses the resulting file while forgetting about grammar inclusion altogether.
That's not because ANTLR is dumb, there are actually Reasons™ for this.
Because of this however, the generated classes corresponding to the composite grammar do not actually inherit (in the OO-sense) from the classes generated for the included grammar. Neither are the latters be aggregated into the formers.

In my case, what I get is a new `TypeCobolParser` class, with as many context-specific internal classes as rules defined my TypeCobol grammar, _PLUS all the rules defined in my imported grammars_.
These are brand new `TypeCobol.Compiler.Parser.Generated.TypeCobolParser.*Context` types ; the majority of them do _exactly the same thing_ of all the context-specific `TypeCobol.Compiler.Parser.Generated.CobolCodeElementsParser.*Context` types, but are **not** related in any way to what already exists.

As you can imagine, I can't make my `TypeCobolBuilder` inherit from or aggregate `CodeElementBuilder` because I can't cast from the `TypeCobol` context objects to the `CobolCodeElements` context objects.

In the short term, it seems

**I'M SCREWED**.

How can I avoid to duplicate all the code related to converting ANTLR contexts into CodeElements ?
The code I'm talking about:
 * represents A LOT of code
 * will suffer many changes as long as we're not ready for production
 * has its inputs a consequence of the grammar ; if you change the grammar, you change the context-specific objects

**Comment by wiztigers** [comment]
Off the top of my head:

**SOLUTION 1**
Convert TypeCobol contexts into CobolCodeElements contexts so `CodeElementBuilder` can be used by `TypeCobolBuilder`.
*comments:*
 * writing this conversion code by hand will be a PITA
 * event if generating the conversion code, what about performance ?

**SOLUTION 2**
Tell ANTLR to re-use objects it already generated for the included grammar.
*comments:*
 * it doesn't seem to be possible
 * I don't have access to the ANTLR reference guide to check if it's possible

**SOLUTION 3**
In a pre-compilation task, modify the ANTLR generated sources to suit my needs. For example:
 * making the context-specific classes NOT internal
 * put them in their own file
 * modify the parsers to use them
*comments:*
 * what about context-specific classes replaced by the TypeCobol grammar ?
 * subject to breaking on changes to the way ANTLR generates code

**Comment by wiztigers** [comment]
@prudholu @osmedile @vavans , any comments?

**Comment by wiztigers** [comment]
Solution implemented in e7702bbc68cbfa8f3e1e38df2697f589627fe0f2, 36f2f865de177b58fd1992003df6a4e602cb6902, cff2762baa016ff3f9acb75d8966ebc1abe33f73.
